ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2/Types.hs

```diff
@@ -0,0 +1,325 @@
+
+data DecisionContext peerAddr objectId object = DecisionContext
+  { dcRng :: StdGen
+  , dcHasObject :: (objectId -> Bool)
+  , dcDecisionPolicy :: DecisionPolicy
+  , dcGlobalState :: DecisionGlobalState peerAddr objectId object
+  , dcPrevDecisions :: Map peerAddr (PeerDecision objectId object)
+  }
+  deriving stock Generic
+  deriving anyclass NFData
+
+-- | Policy for making decisions
+data DecisionPolicy = DecisionPolicy
+  { dpMaxNumObjectIdsReq :: !NumObjectIdsReq
+  -- ^ a maximal number of objectIds requested at once.
+  , dpMaxNumObjectsOutstanding :: !NumObjectsOutstanding
+  -- ^ maximal number of objects in the outstanding FIFO.
+  -- Must be the same as the outbound peer's value.
+  , dpMaxNumObjectsInflightPerPeer :: !NumObjectsReq
+  -- ^ a limit of objects in-flight from a single peer.
+  , dpMaxNumObjectsInflightTotal :: !NumObjectsReq
+  -- ^ a limit of objects in-flight from all peers for this node.
+  , dpTargetObjectRedundancy :: !ObjectMultiplicity
+  -- ^ from how many peers download the `objectId` simultaneously
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData, NoThunks)
+
+-- | Decision made by the decision logic.  Each peer will receive a 'Decision'.
+--
+-- /note:/ it is rather non-standard to represent a choice between requesting
+-- `objectId`s and `object`'s as a product rather than a sum type. The client will
+-- need to download `object`s first and then send a request for more objectIds (and
+-- acknowledge some `objectId`s). Due to pipelining each client will request
+-- decision from the decision logic quite often (every two pipelined requests).
+--
+-- TODO: in the previous design, we prefiltered active peers before calling
+-- `makeDecision`, so that a decision once taken would make the peer non-active
+-- (e.g. it won't be returned by `filterActivePeers`) for longer, and thus the
+-- expensive `makeDecision` computation would not need to take that peer into
+-- account. This is no longer the case, but we could reintroduce this optimization
+-- if needed.
+data PeerDecision objectId object = PeerDecision
+  { pdNumIdsToAck :: !NumObjectIdsAck
+  -- ^ objectId's to acknowledge
+  , pdNumIdsToReq :: !NumObjectIdsReq
+  -- ^ number of objectId's to request
+  , pdCanPipelineIdsRequests :: !Bool
+  -- ^ the object-submission protocol only allows to pipeline `objectId`'s requests
+  -- if we have non-acknowledged `objectId`s.
+  , pdObjectsToReqIds :: !(Set objectId)
+  -- ^ objectId's to download.
+  , pdStatus :: !PeerDecisionStatus
+  -- ^ Whether the peer is actually executing the said decision
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData, NoThunks)
+
+data PeerDecisionStatus
+  = DecisionUnread
+  | DecisionBeingActedUpon
+  | DecisionCompleted
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData, NoThunks)
+
+-- | A placeholder when no decision has been made, at the beginning of a loop.
+-- Nothing should be read from it except its status.
```

Comment:
This feels a bit strange. Would it make sense to inject a `Maybe` somewhere to account for this?

Reply:
This comment has been written when all fields values except `pdStatus` were `error "..."`. Now it's plenty safe to read any field of these decisions.

And I don't think it would make sense to use a `Maybe`. Because `pdStatus` already guard the reading of such decision.

Reply:
After some discussion, see https://github.com/IntersectMBO/ouroboros-consensus/pull/1698#discussion_r2518714097